common/transport/amqp/src/amqp.ts

```diff
@@ -48,70 +123,85 @@ export class Amqp {
     - not reconnect on failure
-    - not reattach sender and receiver links on failure
-    - not reestablish sessions on failure]*/
-    this._amqp = new amqp10.Client(amqp10.Policy.merge(<any>{
-      session: {
-        reestablish: {
-          retries: 0,
-          forever: false
-        }
-      },
-      senderLink: {
-        attach: {
-          maxMessageSize: 0,
-        },
-        encoder: (body: any): any => {
-          if (typeof body === 'string') {
-            return new Buffer(body, 'utf8');
-          } else {
-            return body;
-          }
-        },
-        reattach: {
-          retries: 0,
-          forever: false
-        }
-      },
-      receiverLink: {
-        attach: {
-          maxMessageSize: 0,
-          receiverSettleMode: autoSettleMode,
-        },
-        decoder: (body: any): any => body,
-        reattach: {
-          retries: 0,
-          forever: false
-        }
-      },
-      // reconnections will be handled at the client level, not the transport level.
-      reconnect: {
-        retries: 0,
-        strategy: 'fibonacci',
-        forever: false
-      }
-    }, amqp10.Policy.EventHub));
+    - not reattach sender and receiver links on failure]*/
 
-    const amqpErrorHandler = (err) => {
-      debug('amqp10 client error: ' + err.toString());
+    this._rheaContainer = rheaCreateContainer();
+    this._rheaContainer.on('error', (context: EventContext) => {
+      debug('Container received a error event' + this._getErrorName(context.connection.error));
+    });
+
+    this._rheaContainer.on('azure-iot-amqp-base:error-indicated', (err: AmqpError) => {
+      debug('azure-iot-amqp-base:error-indicated invoked ' + this._getErrorName(err));
       this._fsm.handle('amqpError', err);
-    };
 
-    this._amqp.on('disconnected', () => {
-      // deferring this is necessary because in some instances
-      // the amqp10 library might want to send a close frame - and we don't want
-      // to trigger anything (especially not reconnection) before it has done so.
-      process.nextTick(() => {
-        this._fsm.handle('amqpDisconnected');
-      });
     });
 
+    // //
+    // // This error handler listens for events from THIS transport SenderLinks or ReceiverLinks.
+    // // These error indications would be as a result of an error indication from the associated
+    // // rhea links that are not associated with a specific request.
+    // //
+    // const amqpErrorHandler = (err) => {
+    //   debug('amqpErrorHandler invoked ' + this._getErrorName(err));
+    //   this._fsm.handle('amqpError', err);
+    // };
+
+    const rheaErrorHandler = (context: EventContext) => {
+      debug('rhea error event handler');
+      this._fsm.handle('error', context);
+    };
+
+    const connectionErrorHandler = (context: EventContext) => {
+      debug('connection error event handler');
+      this._fsm.handle('connection_error', context);
```

Comment:
>connection_error [](start = 24, length = 16)

are those the exact names of the events we are receiving? do we need to wrap those in a function or can we just use this._fsm.handle directly as the event handler?

Reply:
I like that idea.  It _seems_ like we could maybe knock out 20 lines of code and a level of indirection that might not be necessary, but it doesn't seem to hurt as is and it gives us a place for breakpoints/additional debug code if necessary.  I think we should leave this as-is unless there's a sound reason to change it besides "I think we could be a little more clever and a little smaller here".  

---
In reply to: [209130807](https://github.com/Azure/azure-iot-sdk-node/pull/344#discussion_r209130807) [](ancestors = 209130807)

Reply:
And I know how much pain Tony went through to make this right, so let's not make him revisit his pain :)

---
In reply to: [209374344](https://github.com/Azure/azure-iot-sdk-node/pull/344#discussion_r209374344) [](ancestors = 209374344,209130807)